In [1]:
import pandas as pd
import numpy as np

from scipy import stats
from scipy.stats import zscore

# Load Dataset
df = pd.read_excel("Trip_Analysis.xlsx")

# Display first 5 rows
print(df.head())

# Dataset Information
print(df.info())

   Trip_Distance  Fare_Amount Ride_Category  Surge_Multiplier  \
0              8   270.569350        Shared               2.5   
1             21   275.161032       Economy               2.5   
2             16   359.748971       Economy               2.0   
3             12   264.044264       Premium               1.5   
4              9   225.256916        Shared               2.0   

   Customer_Rating Ride_Time  
0                3  Non-Peak  
1                3  Non-Peak  
2                5  Non-Peak  
3                5  Non-Peak  
4                2  Non-Peak  
<class 'pandas.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Trip_Distance     200 non-null    int64  
 1   Fare_Amount       200 non-null    float64
 2   Ride_Category     200 non-null    str    
 3   Surge_Multiplier  200 non-null    float64
 4   Customer_Rating   200 non-null    int64  
 

In [2]:
pip install openpyxl

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.0.1 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
pip show openpyxl

Name: openpyxl
Version: 3.1.5
Summary: A Python library to read/write Excel 2010 xlsx/xlsm files
Home-page: https://openpyxl.readthedocs.io
Author: See AUTHORS
Author-email: charlie.clark@clark-consulting.eu
License: MIT
Location: c:\Users\Poornima\AppData\Local\Programs\Python\Python314\Lib\site-packages
Requires: et-xmlfile
Required-by: 
Note: you may need to restart the kernel to use updated packages.


In [4]:
import pandas as pd

df = pd.read_excel("Trip_Analysis.xlsx")

print(df.head())

   Trip_Distance  Fare_Amount Ride_Category  Surge_Multiplier  \
0              8   270.569350        Shared               2.5   
1             21   275.161032       Economy               2.5   
2             16   359.748971       Economy               2.0   
3             12   264.044264       Premium               1.5   
4              9   225.256916        Shared               2.0   

   Customer_Rating Ride_Time  
0                3  Non-Peak  
1                3  Non-Peak  
2                5  Non-Peak  
3                5  Non-Peak  
4                2  Non-Peak  


In [5]:
print("Rows and Columns:", df.shape)
print(df.info())
print(df.isnull().sum())

Rows and Columns: (200, 6)
<class 'pandas.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Trip_Distance     200 non-null    int64  
 1   Fare_Amount       200 non-null    float64
 2   Ride_Category     200 non-null    str    
 3   Surge_Multiplier  200 non-null    float64
 4   Customer_Rating   200 non-null    int64  
 5   Ride_Time         200 non-null    str    
dtypes: float64(2), int64(2), str(2)
memory usage: 9.5 KB
None
Trip_Distance       0
Fare_Amount         0
Ride_Category       0
Surge_Multiplier    0
Customer_Rating     0
Ride_Time           0
dtype: int64


1. Import Libraries and Load Dataset
Explanation
Pandas → Data manipulation
NumPy → Mathematical calculations
SciPy → Statistical tests

In [7]:
import pandas as pd
import numpy as np

from scipy import stats
from scipy.stats import zscore

# Load Dataset
df = pd.read_excel("Trip_Analysis.xlsx")

# Display first 5 rows
print(df.head())

# Dataset Information
print(df.info())

   Trip_Distance  Fare_Amount Ride_Category  Surge_Multiplier  \
0              8   270.569350        Shared               2.5   
1             21   275.161032       Economy               2.5   
2             16   359.748971       Economy               2.0   
3             12   264.044264       Premium               1.5   
4              9   225.256916        Shared               2.0   

   Customer_Rating Ride_Time  
0                3  Non-Peak  
1                3  Non-Peak  
2                5  Non-Peak  
3                5  Non-Peak  
4                2  Non-Peak  
<class 'pandas.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Trip_Distance     200 non-null    int64  
 1   Fare_Amount       200 non-null    float64
 2   Ride_Category     200 non-null    str    
 3   Surge_Multiplier  200 non-null    float64
 4   Customer_Rating   200 non-null    int64  
 

2. Data Cleaning Using Pandas & NumPy

Formula
Mean:

xˉ=∑x
______
	n
Explanation
Missing values are replaced by the average value of the column.

In [9]:
# Check missing values
print(df.isnull().sum())

# Remove duplicates
df = df.drop_duplicates()

# Fill missing numerical values with mean
num_cols = df.select_dtypes(include=np.number).columns

for col in num_cols:
    df[col].fillna(df[col].mean(), inplace=True)

print(df.isnull().sum())

Trip_Distance       0
Fare_Amount         0
Ride_Category       0
Surge_Multiplier    0
Customer_Rating     0
Ride_Time           0
dtype: int64
Trip_Distance       0
Fare_Amount         0
Ride_Category       0
Surge_Multiplier    0
Customer_Rating     0
Ride_Time           0
dtype: int64


C:\Users\Poornima\AppData\Local\Temp\ipykernel_20844\2104024684.py:11: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df[col].fillna(df[col].mean(), inplace=True)


3. Descriptive Statistics
Formulas
Mean

x
ˉ
=
n
∑x​

In [30]:
print(df.describe())

print("Mean")
print(df.mean(numeric_only=True))

print("Median")
print(df.median(numeric_only=True))

print("Mode")
print(df.mode().iloc[0])

print("Variance")
print(df.var(numeric_only=True))

print("Standard Deviation")
print(df.std(numeric_only=True))

print("Skewness")
print(df.skew(numeric_only=True))

print("Kurtosis")
print(df.kurtosis(numeric_only=True))

       Trip_Distance  Fare_Amount  Surge_Multiplier  Customer_Rating  \
count     200.000000   200.000000        200.000000       200.000000   
mean       15.640000   260.312376          1.677500         2.930000   
std         8.433625    78.706127          0.548862         1.444136   
min         2.000000    37.585916          1.000000         1.000000   
25%         8.000000   208.528826          1.000000         2.000000   
50%        15.000000   256.907103          1.500000         3.000000   
75%        23.000000   319.763782          2.000000         4.000000   
max        29.000000   468.753773          2.500000         5.000000   

             zscore  
count  2.000000e+02  
mean  -1.643130e-16  
std    1.002509e+00  
min   -2.836950e+00  
25%   -6.595865e-01  
50%   -4.337423e-02  
75%    7.572548e-01  
max    2.654996e+00  
Mean
Trip_Distance       1.564000e+01
Fare_Amount         2.603124e+02
Surge_Multiplier    1.677500e+00
Customer_Rating     2.930000e+00
zscore          

4. Linear Algebra

Vector

In [31]:
distance = np.array(df['Trip_Distance'])

print(distance[:10])

[ 8 21 16 12  9 22  8 27 20 24]


Matrix

In [32]:
matrix = df[['Trip_Distance',
             'Fare_Amount',
             'Surge_Multiplier']].values

print(matrix[:5])

[[  8.         270.5693496    2.5       ]
 [ 21.         275.16103167   2.5       ]
 [ 16.         359.74897068   2.        ]
 [ 12.         264.04426352   1.5       ]
 [  9.         225.25691628   2.        ]]


Determinant

In [33]:
corr_matrix = df[['Trip_Distance',
                  'Fare_Amount',
                  'Surge_Multiplier']].corr()

det = np.linalg.det(corr_matrix)

print("Determinant =", det)

Determinant = 0.9991304562284103


Eigenvalues and Eigenvectors

In [34]:
eigenvalues, eigenvectors = np.linalg.eig(corr_matrix)

print("Eigenvalues")
print(eigenvalues)

print("Eigenvectors")
print(eigenvectors)

Eigenvalues
[1.03424335 0.98322854 0.98252811]
Eigenvectors
[[ 0.57397825  0.78738672 -0.22488024]
 [ 0.57726178 -0.58384652 -0.57086958]
 [ 0.58079067 -0.19785195  0.78964347]]


Formula
∣A∣
Interpretation
Eigenvalues show importance of factors.
Larger eigenvalue = stronger influence.

5. Probability

Basic Probability
Formula
P(A)=
Number of favorable outcomes/ Total outcomes


In [35]:
prob_economy = len(df[df['Ride_Category']=="Economy"])/len(df)

print(prob_economy)

0.41


Conditional Probability
Formula
P(A|B)={P(AintersetionB)}/{P(B)}

In [36]:
peak_economy = len(df[(df['Ride_Category']=="Economy") &
                      (df['Ride_Time']=="Peak")])

peak = len(df[df['Ride_Time']=="Peak"])

conditional_prob = peak_economy/peak

print(conditional_prob)

0.42718446601941745


Bayes Theorem
Formula
P(A|B)={P(B|A)P(A)}{P(B)}

In [37]:
P_A = prob_economy
P_B_given_A = conditional_prob
P_B = peak/len(df)

bayes = (P_B_given_A * P_A)/P_B

print(bayes)

0.3400886040154585


6. Correlation and Covariance

Pearson Correlation
Formula
r=Cov(X,Y)/σXσy

In [38]:
pearson = df['Trip_Distance'].corr(df['Fare_Amount'])

print(pearson)

0.01681311099331091


Spearman Correlation

In [39]:
spearman = stats.spearmanr(
    df['Surge_Multiplier'],
    df['Customer_Rating']
)

print(spearman)

SignificanceResult(statistic=np.float64(-0.04709083768799108), pvalue=np.float64(0.5078698477562072))


Covariance

In [40]:
covariance = df['Trip_Distance'].cov(df['Fare_Amount'])

print(covariance)

11.160172520076761


7. Outlier Detection
Quartiles & IQR
Formula
IQR=Q3−Q1

In [41]:
Q1 = df['Fare_Amount'].quantile(0.25)
Q3 = df['Fare_Amount'].quantile(0.75)

IQR = Q3 - Q1

lower = Q1 - 1.5*IQR
upper = Q3 + 1.5*IQR

outliers = df[(df['Fare_Amount'] < lower) |
              (df['Fare_Amount'] > upper)]

print(outliers)

    Trip_Distance  Fare_Amount Ride_Category  Surge_Multiplier  \
10             12    37.585916       Economy               1.0   

    Customer_Rating Ride_Time   zscore  
10                5  Non-Peak -2.83695  


Z-score
Formula
z={x-mu}/{sigma}

In [42]:
df['zscore'] = zscore(df['Fare_Amount'])

outliers_z = df[np.abs(df['zscore']) > 3]

print(outliers_z)

Empty DataFrame
Columns: [Trip_Distance, Fare_Amount, Ride_Category, Surge_Multiplier, Customer_Rating, Ride_Time, zscore]
Index: []


8. Inferential Statistics
Formula
CI=xˉ±z(s/root of n)

In [43]:
mean = df['Fare_Amount'].mean()

std = df['Fare_Amount'].std()

n = len(df)

margin = 1.96*(std/np.sqrt(n))

lower = mean - margin
upper = mean + margin

print(lower, upper)

249.40426328850282 271.2204887203723


Bootstrapping
Explanation

Bootstrapping creates many samples and estimates confidence intervals.

In [44]:
boot_means = []

for i in range(1000):
    sample = df['Fare_Amount'].sample(
        frac=1,
        replace=True
    )

    boot_means.append(sample.mean())

print(np.percentile(boot_means,[2.5,97.5]))

[249.60301992 271.12789895]


9. Hypothesis Testing

Q1 Distance vs Fare

In [45]:
r,p = stats.pearsonr(
    df['Trip_Distance'],
    df['Fare_Amount']
)

print(r,p)

0.016813110993310933 0.8132001771984607


Decision

In [46]:
if p < 0.05:
    print("Significant Relationship")
else:
    print("No Significant Relationship")

No Significant Relationship


Q2 Premium vs Economy (T-Test)

In [47]:
premium = df[df['Ride_Category']=="Premium"]['Fare_Amount']

economy = df[df['Ride_Category']=="Economy"]['Fare_Amount']

t,p = stats.ttest_ind(
    premium,
    economy
)

print(t,p)

-0.4074012822247926 0.6843445588292055


Hypothesis

H₀: Mean Premium Fare = Mean Economy Fare

H₁: Mean Premium Fare ≠ Mean Economy Fare

Q3 Chi-Square Test

In [48]:
economy = df[df['Ride_Category']=="Economy"]['Fare_Amount']

premium = df[df['Ride_Category']=="Premium"]['Fare_Amount']

shared = df[df['Ride_Category']=="Shared"]['Fare_Amount']

f,p = stats.f_oneway(
    economy,
    premium,
    shared
)

print(f,p)

0.9284988792342401 0.39686861659468403


Q4 ANOVA

In [49]:
economy = df[df['Ride_Category']=="Economy"]['Fare_Amount']

premium = df[df['Ride_Category']=="Premium"]['Fare_Amount']

shared = df[df['Ride_Category']=="Shared"]['Fare_Amount']

f,p = stats.f_oneway(
    economy,
    premium,
    shared
)

print(f,p)

0.9284988792342401 0.39686861659468403


Question 1: Does Trip Distance Affect Fare Amount?
Pearson Correlation
| Correlation | Meaning     |
| ----------- | ----------- |
| 0           | No relation |
| 0.1 - 0.3   | Weak        |
| 0.3 - 0.7   | Moderate    |
| >0.7        | Strong      |


In [50]:
pearson = df['Trip_Distance'].corr(df['Fare_Amount'])

print("Pearson Correlation:", pearson)

Pearson Correlation: 0.01681311099331091


In [51]:
covariance = df['Trip_Distance'].cov(df['Fare_Amount'])

print("Covariance:", covariance)

Covariance: 11.160172520076761


Question: Does trip distance significantly affect fare amount?

Answer:

No. The Pearson correlation coefficient is 0.0168, which is very close to zero. This indicates that trip distance has a negligible effect on fare amount in the given dataset.

Question 2: Premium vs Economy Fare
Descriptive Statistics

In [52]:
premium = df[df['Ride_Category']=="Premium"]['Fare_Amount']

economy = df[df['Ride_Category']=="Economy"]['Fare_Amount']

print("Premium Mean:", premium.mean())
print("Economy Mean:", economy.mean())

Premium Mean: 252.45503727502648
Economy Mean: 257.74844283796944


T-Test

In [53]:
from scipy.stats import ttest_ind

t_stat,p_value = ttest_ind(
    premium,
    economy,
    equal_var=False
)

print("T-statistic:",t_stat)
print("P-value:",p_value)

T-statistic: -0.40600606266641986
P-value: 0.6854530831483112


Decision

In [54]:
if p_value < 0.05:
    print("Significant Difference")
else:
    print("No Significant Difference")

No Significant Difference


Question: Are Premium rides significantly more expensive than Economy rides?

Answer:

No. The t-test produced a p-value of 0.6855, which is greater than 0.05. Therefore, there is no statistically significant difference between Premium and Economy ride fares.

Question 3: Surge Pricing vs Customer Ratings

Spearman Correlation

In [55]:
from scipy.stats import spearmanr

corr,p = spearmanr(
    df['Surge_Multiplier'],
    df['Customer_Rating']
)

print(corr)
print(p)

-0.04709083768799108
0.5078698477562072


No. The Spearman correlation coefficient is -0.0471, indicating a very weak negative relationship, and the p-value (0.5079) is greater than 0.05. Therefore, surge pricing does not significantly affect customer ratings.

Question 4: Peak vs Non-Peak Revenue

In [56]:
peak = df[df['Ride_Time']=="Peak"]['Fare_Amount']

non_peak = df[df['Ride_Time']=="Non-Peak"]['Fare_Amount']

from scipy.stats import ttest_ind

t,p = ttest_ind(
    peak,
    non_peak,
    equal_var=False
)

print(t,p)

-0.596491792388811 0.5515482990013166


No. The p-value is 0.5515, which is greater than 0.05, indicating that the difference is not statistically significant.

Question 5: Outlier Detection

Quartiles

In [57]:
Q1 = df['Fare_Amount'].quantile(0.25)

Q3 = df['Fare_Amount'].quantile(0.75)

IQR = Q3-Q1

print("Q1 =",Q1)
print("Q3 =",Q3)
print("IQR =",IQR)

Q1 = 208.5288261048614
Q3 = 319.7637824819526
IQR = 111.23495637709121


In [ ]:
peak = df[df['Ride_Time']=="Peak"]['Fare_Amount']

non_peak = df[df['Ride_Time']=="Non-Peak"]['Fare_Amount']

from scipy.stats import ttest_ind

t,p = ttest_ind(
    peak,
    non_peak,
    equal_var=False
)

print(t,p)

-0.596491792388811 0.5515482990013166


IQR Method

In [58]:
lower = Q1 - 1.5*IQR
upper = Q3 + 1.5*IQR

outliers = df[
    (df['Fare_Amount']<lower) |
    (df['Fare_Amount']>upper)
]

print(outliers)

    Trip_Distance  Fare_Amount Ride_Category  Surge_Multiplier  \
10             12    37.585916       Economy               1.0   

    Customer_Rating Ride_Time   zscore  
10                5  Non-Peak -2.83695  


Z-Score

In [59]:
from scipy.stats import zscore

df['Zscore'] = zscore(df['Fare_Amount'])

outliers_z = df[
    abs(df['Zscore'])>3
]

print(outliers_z)

Empty DataFrame
Columns: [Trip_Distance, Fare_Amount, Ride_Category, Surge_Multiplier, Customer_Rating, Ride_Time, zscore, Zscore]
Index: []


One outlier was detected using the IQR method, while no outliers were detected using the Z-score method.

Question 6: Probability

Ride Demand Probability

In [60]:
prob = df['Ride_Category'].value_counts(normalize=True)

print(prob)

Ride_Category
Economy    0.41
Shared     0.30
Premium    0.29
Name: proportion, dtype: float64


Conditional Probability

Probability of Economy Ride During Peak Hours

In [61]:
economy_peak = len(
    df[
        (df['Ride_Category']=="Economy")
        &
        (df['Ride_Time']=="Peak")
    ]
)

peak = len(
    df[df['Ride_Time']=="Peak"]
)

print(economy_peak/peak)

0.42718446601941745


Question: What is the probability of an Economy ride during Peak hours?

Answer:

The conditional probability is 0.4272, meaning there is approximately a 42.72% chance that a Peak-hour ride belongs to the Economy category.

Question 7: Eigenvalues Analysis

In [63]:
import numpy as np

numeric_data = df[
[
'Trip_Distance',
'Fare_Amount',
'Surge_Multiplier',
'Customer_Rating'
]
]

corr_matrix = numeric_data.corr()

eigenvalues,eigenvectors = np.linalg.eig(
    corr_matrix
)

print(eigenvalues)
print(eigenvectors)

[0.88618284 1.1055071  0.98266344 1.02564662]
[[-0.53777588  0.58844968  0.5492467   0.25070326]
 [ 0.43248566 -0.20303316  0.25762521  0.83985888]
 [-0.2337979   0.33830171 -0.79495255  0.44602795]
 [-0.68490276 -0.70573115  0.00278261  0.18122917]]


Question: Why are eigenvalues used in data analysis?

Answer:

Eigenvalues help identify the most important patterns in data. A larger eigenvalue indicates a component that explains more variation and has greater influence on the dataset.

Question 8: Confidence Interval

In [64]:
import numpy as np

mean = df['Fare_Amount'].mean()

std = df['Fare_Amount'].std()

n = len(df)

margin = 1.96*(std/np.sqrt(n))

lower = mean-margin

upper = mean+margin

print(lower)
print(upper)

249.40426328850282
271.2204887203723


Bootstrapping

In [65]:
boot_means = []

for i in range(1000):

    sample = df['Fare_Amount'].sample(
        frac=1,
        replace=True
    )

    boot_means.append(
        sample.mean()
    )

print(
    np.percentile(
        boot_means,
        [2.5,97.5]
    )
)

[249.65381245 270.58128404]


The bootstrap confidence interval is very close to the traditional confidence interval, indicating that the estimate of the average fare is stable and reliable.

Final Conclusions (All Questions)
Q1. Distance vs Fare Amount

Pearson correlation = 0.0168 and covariance = 11.16. The relationship between Trip Distance and Fare Amount is extremely weak. Distance does not significantly influence fare amount.

Q2. Premium vs Economy Fare

The average Premium fare was 252.46, while the Economy fare was 257.75. The t-test produced a p-value of 0.6855, indicating no significant difference between the two ride categories.

Q3. Surge Pricing vs Customer Ratings

Spearman correlation = -0.0471 and p-value = 0.5079. Surge pricing has no statistically significant impact on customer ratings.

Q4. Peak vs Non-Peak Revenue

The t-test produced a p-value of 0.5515. Revenue during Peak and Non-Peak periods does not differ significantly.

Q5. Outlier Detection

One outlier was detected using the IQR method. No outliers were identified using the Z-score method. The dataset contains very few anomalous fare values.

Q6. Probability Analysis

The probability of an Economy ride during Peak hours was 42.72%, indicating strong demand for Economy rides during busy periods.

Q7. Eigenvalue Analysis

The largest eigenvalue was 1.1055. No single operational factor dominates business performance; multiple factors contribute collectively.

Q8. Confidence Intervals & Bootstrapping

The mean fare estimate is statistically reliable, with confidence intervals consistently ranging between approximately 250 and 271.